rfcs/syntax-safe-navigation-operator.md

```diff
@@ -0,0 +1,128 @@
+# Safe navigation postfix operator (?)
+
+## Summary
+
+Introduce syntax to navigate through `nil` values, or short-circuit with `nil` if it was encountered.
+
+
+## Motivation
+
+nil values are very common in Lua, and take care to prevent runtime errors. 
+
+Currently, attempting to index `dog.name` while caring for `dog` being nil requires some form of the following:
+
+```lua
+local dogName = nil
+if dog ~= nil then
+    dogName = dog.name
+end
+```
+
+...or the unusual to read...
+
+```lua
+local dogName = dog and dog.name
+```
+
+...which will return `false` if `dog` is `false`, instead of throwing an error because of the index of `false.name`.
+
+This trick gets worse in the case of calling methods. For example, let's suppose we wanted to call `dog.fetch()`, while `dog` is still potentially nil.
+
+The one-line `and` trick will no longer work (or will get less readable as you try to shoe-horn this in), so we must:
+
+```lua
+if dog ~= nil then
+    dog.fetch()
+end
+```
+
+But this gets even worse when it comes to chained indexing. Let's suppose we wanted to run `dog.owner.handshake()`, while `dog` can be nil and `owner` can be nil.
+
+```lua
+if dog ~= nil and dog.owner ~= nil then
+    dog.owner.handshake()
+end
+```
+
+...which gets even worse in the context of calling this function in, say, another function:
+
+```lua
+-- Oops! dog and dog.owner can be nil
+logDogName(getLogger(), dog.name, dog.owner:getDisplayName())
+
+-- In order to preserve this order (assuming argument execution order mattered)...
+local logger = getLogger()
+local name = dog and dog.name
+local displayName
+
+if dog.owner ~= nil then
+    displayName = dog.owner:getDisplayName()
+end
+
+logDogName(logger, name, displayName)
+```
+
+## Design
+
+The safe navigation operator will make all of these smooth, by supporting `x?.y` and similar indexing operators. `dog?.name` would resolve to `nil` if `dog` was nil, or the name otherwise. `owner?.handshake()` would only call `handshake` if `owner` is not nil.
+
+The long example would turn into:
+
+```lua
+logDogName(getLogger(), dog?.name, dog?.owner?:getDisplayName())
+```
+
+Failing the nil-safety check early would make the entire expression nil, for instance `dog?.body.legs` would resolve to `nil` if `dog` is nil, rather than resolve `dog?.body` into nil, then turning into `nil.legs`.
+
+The list of valid operators to follow the safe navigation operator would be:
+
+```lua
+dog?.name --[[ is the same as ]] if dog == nil then nil else dog.name

```

Comment:
Is the intent actually for this to only work with nil values, rather than a general falsiness check? It has odd interactions with operators like `and` if not.

Reply:
Yeah, this is only for nil. I think that attempting to index a false value *should* error. I mention the falsiness thing in the `dog and dog.name` mention of the RFC.